test/hybrid_test.py

```diff
@@ -45,99 +46,128 @@ def s_expr(x):
+    solver_highd = icepack.solvers.FlowSolver(model_highd, **opts)
+    u_highd = solver_highd.diagnostic_solve(
         velocity=u0,
         thickness=h,
         surface=s,
         fluidity=A,
         friction=C
     )
 
-    U2D, U3D = u2d.dat.data_ro, u3d.dat.data_ro
-    assert np.linalg.norm(U3D - U2D) / np.linalg.norm(U2D) < 1e-2
+    U_lowD, U_highD = u_lowd.dat.data_ro, u_highd.dat.data_ro
+    assert np.linalg.norm(U_highD - U_lowD) / np.linalg.norm(U_lowD) < 1e-2
+
+
+@pytest.mark.parametrize('dim', ['xz', 'xyz'])
+def test_diagnostic_solver(dim):
+    Nx, Ny, Nz = 32, 32, 32
+    if dim == 'xz':
+        opts = {'dirichlet_ids': [1], 'tol': 1e-12}
+        mesh_lowd = firedrake.IntervalMesh(Nx, Lx)
+        mesh = firedrake.ExtrudedMesh(mesh_lowd, layers=1)
+        x, ζ = firedrake.SpatialCoordinate(mesh)
+        u_expr = (0.95 + 0.05 * ζ) * exact_u(x)
+        xs = np.array([(Lx/2, k / Nz) for k in range(Nz + 1)])
+    elif dim == 'xyz':
+        opts = {'dirichlet_ids': [1, 3, 4], 'tol': 1e-12}
+        mesh_lowd = firedrake.RectangleMesh(Nx, Ny, Lx, Ly)
+        mesh = firedrake.ExtrudedMesh(mesh_lowd, layers=1)
+        x, y, ζ = firedrake.SpatialCoordinate(mesh)
+        u_expr = firedrake.as_vector(((0.95 + 0.05 * ζ) * exact_u(x), 0))
+        xs = np.array([(Lx/2, Ly/2, k / Nz) for k in range(Nz + 1)])
 
-
-def test_diagnostic_solver():
-    Nx, Ny = 32, 32
-    mesh2d = firedrake.RectangleMesh(Nx, Ny, Lx, Ly)
-    mesh = firedrake.ExtrudedMesh(mesh2d, layers=1)
     Q = firedrake.FunctionSpace(
         mesh, family='CG', degree=2, vfamily='DG', vdegree=0)
 
-    x, y, ζ = firedrake.SpatialCoordinate(mesh)
     h = firedrake.interpolate(h0 - dh * x / Lx, Q)
     s = firedrake.interpolate(d + h0 - dh + ds * (1 - x / Lx), Q)
-    u_expr = firedrake.as_vector(((0.95 + 0.05 * ζ) * exact_u(x), 0))
 
     A = firedrake.Constant(icepack.rate_factor(254.15))
     C = firedrake.Constant(0.001)
 
     model = icepack.models.HybridModel()
-    opts = {'dirichlet_ids': [1, 3, 4], 'tol': 1e-12}
 
     max_degree = 5
-    Nz = 32
-    xs = np.array([(Lx/2, Ly/2, k / Nz) for k in range(Nz + 1)])
     us = np.zeros((max_degree + 1, Nz + 1))
     for vdegree in range(max_degree, 0, -1):
         solver = icepack.solvers.FlowSolver(model, **opts)
-        V = firedrake.VectorFunctionSpace(
-            mesh, dim=2, family='CG', degree=2, vfamily='GL', vdegree=vdegree
-        )
+        if dim == 'xz':
+            V = firedrake.FunctionSpace(
+                mesh, family='CG', degree=2, vfamily='DG', vdegree=vdegree)
+            V0 = firedrake.FunctionSpace(
+                mesh, family='CG', degree=2, vfamily='DG', vdegree=0)
+        elif dim == 'xyz':
+            V = firedrake.VectorFunctionSpace(
+                mesh, dim=2, family='CG', degree=2, vfamily='DG', vdegree=vdegree)
+            V0 = firedrake.VectorFunctionSpace(
+                mesh, dim=2, family='CG', degree=2, vfamily='DG', vdegree=0)
+
```

Comment:
Outside of the loop, you could instead do something like this:

    fn_space_kwargs = {'family': 'CG', 'degree': 2, 'vfamily': 'DG'}
    if dim == 'xyz':
        fn_space_kwargs['dim'] = 2

Then inside the loop you can just do

    V = firedrake.VectorFunctionSpace(mesh, vdegree=vdegree, **fn_space_kwargs)
    V0 = firedrake.VectorFunctionSpace(mesh, vdegree=0, **fn_space_kwargs)

with no conditionals.

Reply:
I tried this, but am getting a failed test:
"FAILED hybrid_test.py::test_diagnostic_solver[xz] - RuntimeError: Expression of length 2 required, got length 1"

Is it allowed to be a vector space if it is only the extruded 1-d mesh?

Reply:
No you're right, I didn't read it carefully enough. We can refactor this later.